[v1.13] Delete IP Label metadata on delete from ipcache

meyskens opened the pull request:
v1.13 backport of https://github.com/cilium/cilium/pull/26958 

```release-note
Delete IP Label metadata on delete from ipcache
```

christarazi reviewed (changes requested):
https://github.com/cilium/cilium/pull/27010#issuecomment-1652559801

christarazi commented on pkg/ipcache/ipcache.go:
```diff
@@ -417,9 +417,10 @@ func (ipc *IPCache) UpsertMetadata(prefix netip.Prefix, src source.Source, resou
 	ipc.metadata.Lock()
-	ipc.metadata.remove(prefix, resource, aux...)
+	ipc.metadata.removeAllForPrefix(prefix)
 	ipc.metadata.Unlock()
```
This will actually break the intended behavior of the metadata resolution. Essentially, there should be an equal number of upserts to removals. In other words, for every upsert-metadata, there is a remove-metadata. This breaks that assumption, which will remove metadata from a prefix prematurely.

christarazi commented on pkg/ipcache/ipcache.go:
```diff
@@ -436,10 +437,6 @@ func (ipc *IPCache) UpsertLabels(prefix netip.Prefix, lbls labels.Labels, src so
 }
 
-func (ipc *IPCache) RemoveLabels(cidr netip.Prefix, lbls labels.Labels, resource ipcacheTypes.ResourceID) {
-	ipc.RemoveMetadata(cidr, resource, lbls)
```
I don't think we should do this because it is a public API which may get used in future versions of v1.13.

christarazi reviewed (changes requested):
My previous review comment still applies.

christarazi commented on pkg/node/manager/manager.go:
```diff
@@ -651,15 +652,9 @@ func (m *Manager) NodeDeleted(n nodeTypes.Node) {
-			m.ipcache.Delete(address.String(), n.Source)
-		}
+		resource := ipcacheTypes.NewResourceID(ipcacheTypes.ResourceKindNode, "", n.Name)
+		m.ipcache.RemoveLabels(prefix, labels.Labels{}, resource)
```
Strictly only this change will cause other bugs because we should be calling `RemoveLabels` on the inverse of wherever `upsertIntoIDMD` is called. In other words, `upsertIntoIDMD` and `RemoveLabels` need to be symmetric.

We already have a convenient function analogous to `upsertIntoIDMD` called `deleteIPCache`. I think the call to `RemoveLabels` should go inside `deleteIPCache` without removing the existing ipcache calls.

meyskens committed:
node: Remove labels from ipcache on delete events

This commit is based on v1.12 changes made in
067523f39bc6a65e766be6602f405246ddf99310
It has the same added function calls but refactored to work with the
resource ID based metadata cache of v1.13.

Previously the node manager code only ever performed upserts into the
ipcache for remote-node labels on node creation/update events. However,
it never cleaned up these labels upon delete. As a result, if a node was
ever removed, then Cilium would continue to consider traffic towards
that IP as traffic reaching towards a remote node. In extreme cases,
this IP could be reallocated to another host in the network which could
cause connectivity disruption to that peer, particularly if the new peer
is subject to CIDR or Entities based policies.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -651,15 +681,12 @@ func (m *Manager) NodeDeleted(n nodeTypes.Node) {
 			prefix = ip.IPToNetPrefix(address.IP.To16())
 		}
 		m.ipcache.Delete(prefix.String(), n.Source)
-	}
 
-	for _, address := range []net.IP{
-		entry.node.IPv4HealthIP, entry.node.IPv6HealthIP,
-		entry.node.IPv4IngressIP, entry.node.IPv6IngressIP,
```
When the node gets deleted, this code still needs to remove the association for the old health and ingress IPs. Was it intentional to remove the iteration here for these additional addresses?

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -651,15 +681,12 @@ func (m *Manager) NodeDeleted(n nodeTypes.Node) {
+		resource := ipcacheTypes.NewResourceID(ipcacheTypes.ResourceKindNode, "", n.Name)
+
+		m.removeFromIDMD(prefix, remoteHostIdentity, resource)
+		m.ipcache.Delete(ip, n.Source)
```
This seems to be using the non-prefix representation of the string like `192.0.2.3`, as compared with the prefix representation like `192.0.2.3/32` as like the call six lines up. However, `NodeUpdated()` doesn't seem to do an `ipcache.Upsert(...)` with this representation. So this seems to be deleting an association in the ipcache that was never created when the node was initially discovered? I don't think that this is correct.
